flang/lib/Lower/OpenMP/OpenMP.cpp

```diff
@@ -521,7 +521,7 @@ struct OpWithBodyGenInfo {
 /// \param [in]   op - the operation the body belongs to.
 /// \param [in] info - options controlling code-gen for the construction.
 template <typename Op>
-static void createBodyOfOp(Op &op, OpWithBodyGenInfo &info) {
+static void createBodyOfOp(mlir::Operation &op, OpWithBodyGenInfo &info) {
```

Comment:
The new version doesn't take any parameters or return any values with the type in the template arguments, so there is no compelling reason to keep it as a template, expecially given its size.

I suggest converting it to a non-template, with an extra parameter "isLoop" passed from outside.  We can have "is_loop" trait to make it easy.

Reply:
That's a good point, thanks for the suggestion. Since it uses the template argument to check for loop operations and also for `omp.parallel`, I thought an enum added to the `OpWithBodyGenInfo` structure could work well in this case. Let me know what you think about this solution.

I'd rather try to avoid queries on `op` here, like checking op traits, calling `isa()` or anything like this, because when lowering loops `op` is the `omp.loop_nest` rather than any wrapper operations.